### ISSUE TITLE
C-API Enhancements

### ISSUE THREAD

**Issue by zenna** [dreal/dreal3#100]
@soonhokong @scungao : Here is a very rough outline of the enhancements I would like for the API.  The following is psuedo-code to describe my intent, but there may be other ways to do it better.  Either way, let's discuss.

- I need access to the CNF structure (and possibly the mapping between the equations and literals) in a way that will allow me to use an external solver to get a model
- I need to be able to go from a boolean model to the equivalent real valued problem that will be sent to the theory solver.
- I need to be able to contract a box from an initial box
- I need to be able to add and remove clauses from the CNF
- When calling the boolean solver many times, I may want it to remember all the clauses it has learned over times (I imagine you are doing this now in dReal)

```C
int main(int argc, char * argv[]) {
    opensmt_init();
    opensmt_context ctx = opensmt_mk_context(qf_nra);
    opensmt_set_precision (ctx, 0.001);
    opensmt_expr x = opensmt_mk_real_var(ctx, "w1" , -3.141592, 3.141592);
    opensmt_expr y = opensmt_mk_real_var(ctx, "w2" , -3.141592, 3.141592);
    opensmt_expr eq1 = opensmt_mk_gt(ctx, x, y);
    opensmt_expr eq2 = opensmt_mk_lt(ctx, y, 0.5);
    opensmt_expr eq3 = opensmt_mk_or(ctx, eq1, eq2);

    // Get the boolean structure in CNF
    bool_vars = api_get_bool_variables(ctx)
    cnf = api_get_cnf_structure(ctx)

    // Check whether the boolean structure is satisfiable
    bool is_sat = api_check(cnf)
    if (!is_sat) {
        return -1;
    }

    model = api_model(cnf) // Get a model from the CNF
    initial_box = api_get_initial_box(ctx)
    theory_conjunct = api_get_theory_from_boolean_model(ctx,model)
    new_box = contract(ctx, theory_conjunct, initial_box)

    // Create a clause from the variables
    clause = myfunc(bool_vars) 
    
    // Add a clause to the CNF
    cnf2 = add_clause(cnf, clause)
    remove_clause(cnf, clause)

}
```

**Comment by scungao** [comment]
Ok I guess this is gonna require some work. :)

Maybe at this point a good thing to do is to open up the different
components in dReal anyway, so that others can easily do their own thing on
top of it. For instance I wanna add some contractors that implement various
optimization algorithms.

A good example for writing about stuff like this is the minisat paper

http://www.decision-procedures.org/handouts/MiniSat.pdf

Basically we would explain all the major steps in the solving process, and
describe the code and how to make modifications for each of them. Namely, a
"Programmer's Manual". There should be plenty of places to send this to.

@soonhokong, @zenna, @nnarodytska, @danbryce, I guess we could all
collaborate for this write up. I'll start a repo for this.

**Comment by soonhokong** [comment]
I agree that it requires some work.

> "4. I need to be able to add and remove clauses from the CNF"

This is already possible by using `opensmt_push` and `opensmt_pop`. Anything asserted after `opensmt_push` will be poped with `opensmt_pop` call. Here is [an example](https://github.com/dreal/dreal3/blob/master/src/tests/c_api/example_03.c) using them.

> "3. I need to be able to contract a box from an initial box"

This one is not there yet, but relatively easier to implement. We need to expose `box` data structure that we're using internally in dReal3. It's written in C++, so we need to provide an equivalent one in C just like the current C API is doing.

**Comment by soonhokong** [comment]
The other items (1,2,5) look more complicated. 

Simply put, dReal is an instantiation of the DPLL<T> framework which consists of a SAT solver and theory solvers. So far, we've only exposed the SMT-level functions such as `assert`/`push`/`pop`/`check` to users via the current C APIs. Now, we're talking about exposing the internal functions/states of the SAT solver and theory solvers in dReal3. Simply exposing them should be that difficult. The hard part is to maintain the invariant of the solver throughout the use of those APIs.

> 1. I need access to the CNF structure (and possibly the mapping between the equations and literals) in a way that will allow me to use an external solver to get a model

> 2. I need to be able to go from a boolean model to the equivalent real valued problem that will be sent to the theory solver.

If you just want to get a formula from dReal, run an external solver, and do something with the result, then it's not difficult to do. However, if you want to send the result back to dReal and resume its solving process, this is another problem. Essentially, it means that you want to add a custom theory solver in dReal3. This is certainly possible and where DPLL<T> shines. But I think you want to implement this theory solver in Julia and want to plug in to dReal via C API, right? If so, I need to think about how to do this.

Please let me know if I understand anything incorrectly.

> "5. When calling the boolean solver many times, I may want it to remember all the clauses it has learned over times (I imagine you are doing this now in dReal)"

Similarly, simply passing the information shouldn't be hard. But if you want to pass some new information back to the SAT solver in dReal.. This is another problem.

**Comment by soonhokong** [comment]
The pseudo code is certainly useful. But can you share with us the big picture that you have where dReal is part of it? I think it helps a lot to clarify what we need to add.

**Comment by scungao** [comment]
@zenna is using dReal to do probabilistic reasoning, and these functions
are indeed what he needs.

Zenna, are you familiar with C++? Perhaps you'll eventually need to start
looking at the dReal code. We could give you walk through of the code base
at some point. Soonho, let's start writing something for that so that
others can look at too.

**Comment by zenna** [comment]
Thanks for your comments.  @soonhokong, @scungao is right in that I am doing probabilistic inference.

At a high level, I will have a model with variables `x1,x2,..,xn`, but there is some core subset of these variables (e.g. `x1` and `x2`) such that values of all the other variables are fully determined by (i.e. functions of) the core variables. If `X` is the projection of the satisfiable region onto the core variables space, I need to sample uniformly from `X`.

To sample uniformly I need to have control over any "choices" at both the theory and Boolean level.  This means I need to control i) after splitting a hyperbox, which subboxes are chosen in what order ii) which boolean model is chosen.

> If you just want to get a formula from dReal, run an external solver, and do something with the result, then it's not difficult to do. However, if you want to send the result back to dReal and resume its solving process, this is another problem

I do not think I need an extra theory solver. As I understand it, for a given boolean model, dReal will branch and prune at the theory level to determine whether it is SAT or UNSAT.  If it is UNSAT it will ask the SAT solver for another model and repeat.

What I require is the ability to control the boolean model selection part.  That is
- dReal will give me the CNF structure
- I will choose a boolean model assuming there is one
- dReal will be able to a) Do the theory solving to SAT or UNSAT with respect to the boolean mode li have chosen
b) just contract with respect to the boolean model I have chosen to a new box from an initial box

Presumably this code is already there for this in some form.

@scungao Yes I know C++ well. That's a good idea.
